zio-kafka/src/main/scala/zio/kafka/consumer/Consumer.scala

```diff
@@ -404,24 +469,163 @@ private[consumer] final class ConsumerLive private[consumer] (
     keyDeserializer: Deserializer[R, K],
     valueDeserializer: Deserializer[R, V],
-    bufferSize: Int
+    bufferSize: Int = 4
   ): ZStream[R, Throwable, CommittableRecord[K, V]] =
     partitionedStream(subscription, keyDeserializer, valueDeserializer).flatMapPar(
       n = Int.MaxValue,
       bufferSize = bufferSize
     )(_._2)
 
+  override def withPlainStream[R, K, V](
+    subscription: Subscription,
+    keyDeserializer: Deserializer[R, K],
+    valueDeserializer: Deserializer[R, V],
+    bufferSize: Int = 4,
+    shutdownTimeout: Duration = 15.seconds
+  )(
+    withStream: ZStream[R, Throwable, CommittableRecord[K, V]] => ZIO[R, Throwable, Any]
+  ): ZIO[R, Throwable, Any] =
+    withPartitionedStream(subscription, keyDeserializer, valueDeserializer, shutdownTimeout) { partitionedStream =>
+      withStream(
+        partitionedStream.flatMapPar(n = Int.MaxValue, bufferSize = bufferSize)(_._2)
+      )
+    }
+
+  /**
+   * Takes a SubscriptionStreamControl for some stream and runs the given ZIO workflow on that stream such that, when
+   * interrupted, stops fetching records and gracefully waits for the ZIO workflow to complete.
+   *
+   * @param streamControl
+   *   Result of `withPlainStream`, `withPartitionedStream` or `withPartitionedAssignmentStream`
+   * @param shutdownTimeout
+   *   Timeout for the workflow to complete after initiating the graceful shutdown
+   * @param withStream
+   *   Takes the stream as input and returns a ZIO workflow that processes the stream. As in most programs the given
+   *   workflow runs until an external interruption, the result value (Any type) is meaningless. `withStream` is
+   *   typically something like `stream => stream.mapZIO(record => ZIO.debug(record)).mapZIO(_.offset.commit)`
+   */
+  private def runWithGracefulShutdown[StreamType <: ZStream[_, _, _], R, E](
+    streamControl: ZIO[Scope, E, SubscriptionStreamControl[StreamType]],
+    shutdownTimeout: Duration
+  )(
+    withStream: StreamType => ZIO[R, E, Any]
+  ): ZIO[R, E, Any] =
+    ZIO.scoped[R] {
+      for {
+        control <- streamControl
+        fib <-
+          withStream(control.stream)
+            .onInterrupt(
+              ZIO.logError("withStream in runWithGracefulShutdown was interrupted, this should not happen")
+            )
+            .tapErrorCause(cause => ZIO.logErrorCause("Error in withStream fiber in runWithGracefulShutdown", cause))
+            .forkScoped
+        result <-
+          fib.join.onInterrupt(
+            control.stop *>
+              fib.join
+                .timeout(shutdownTimeout)
+                .someOrElseZIO(
+                  ZIO.logError(
+                    "Timeout joining withStream fiber in runWithGracefulShutdown. Not all pending commits may have been processed."
+                  )
+                )
+                .tapErrorCause(cause =>
+                  ZIO.logErrorCause("Error joining withStream fiber in runWithGracefulShutdown", cause)
```

Comment:
If I understand this correctly, the problem is not that the join fails, but more likely the stream itself.
```suggestion
                  ZIO.logErrorCause("Stream failed while awaiting its graceful shutdown", cause)
```